### PR TITLE
AST_to_IL: Fix translation of side-effects in Bool exprs

### LINKED ISSUE #7199: Constant propagation fails to consider branches and conditionals
**Describe the bug**

In what follows, both methods are functionally the same.

```java
class A {
    
    public void method(boolean condition){
      var str = "hello";
      if(condition){
        if("goodbye".equals(str = "goodbye"))
	  X.test(str);
      }
      // ruleid: untitled_rule
      X.test(str);
    }

    public void method2(boolean condition){
      var str = "hello";
      if(condition && "goodbye".equals(str = "goodbye")){
        X.test(str);
      }
      // ruleid: untitled_rule
      X.test(str);
    }
}
```

~~- The pattern `X.test("hello")` fails to match the second call in both methods. It is possible if `condition` is `false`.~~
~~- The pattern `X.test("goodbye")` fails to match the second call in the first method. It is possible if `condition` is `true`.~~
See @IagoAbal's comment (https://github.com/returntocorp/semgrep/issues/7199#issuecomment-1440347945) below. The behavior for `X.test("hello")` is correct but still wrong for `X.test("goodbye")`, as it should not match the second call on `method2`.

**To Reproduce**
[https://semgrep.dev/s/elDE](https://semgrep.dev/s/elDE)

**Expected behavior**
~~`X.test("hello")` should match the second `X.test(str)` call and `X.test("goodbye")` should match all calls.~~
`X.test("goodbye")` should not match the second call on `method2`.

**What is the priority of the bug to you?**

- [ ] P0: blocking your adoption of Semgrep or workflow
- [ ] P1: important to fix or quite annoying
- [X] P2: regular bug that should get fixed

### PR DESCRIPTION
**Pull request by IagoAbal**
We were interpreting `A && x = 42` as `x = 42; A && x`, but the side-effects of `x = 42` only take effect if `A` holds.

We fix this by guarding these side-effects, e.g. interpreting `A && x = 42` as `if (A) x = 42; A && x`.
    
Unfortunately, this will have some side-effects on sym-prop, because we kill sym-prop values at JOIN points, so any sym-prop values are lost after an if-statement. Sym-prop is experimental so that's OK.

Closes #7199

test plan:
make test # tests added

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: cli/tests/e2e/snapshots/test_dependency_aware_rules/test_dependency_aware_rules/rulesdependency_awarelog4shell.yaml-dependency_awarelog4j/results.txt [hunk 1]
```diff
@@ -39,19 +39,6 @@ SEMGREP_USER_AGENT_APPEND="pytest" SEMGREP_SETTINGS_FILE="<MASKED>" SEMGREP_VERS
               "line": 33,
               "offset": 1035
             },
-            "propagated_value": {
```
Comment (aryx) [inline cli/tests/e2e/snapshots/test_dependency_aware_rules/test_dependency_aware_rules/rulesdependency_awarelog4shell.yaml-dependency_awarelog4j/results.txt:42]: this seems unrelated. cc @mmcqd

File: src/analyzing/AST_to_IL.ml [hunk 2]
```diff
@@ -445,8 +453,14 @@ and assign env lhs tok rhs_exp e_gen =
 and expr_aux env ?(void = false) e_gen =
```
Comment (aryx) [inline src/analyzing/AST_to_IL.ml:456]: could you explain what this void parameter is?

**Review by aryx — APPROVED** [review]

File: cli/tests/e2e/snapshots/test_dependency_aware_rules/test_dependency_aware_rules/rulesdependency_awarelog4shell.yaml-dependency_awarelog4j/results.txt (hunk 1 above)
Comment (IagoAbal) [reply cli/tests/e2e/snapshots/test_dependency_aware_rules/test_dependency_aware_rules/rulesdependency_awarelog4shell.yaml-dependency_awarelog4j/results.txt:42]: It is related to this change but unrelated to SSC:

> Unfortunately, this will have some side-effects on sym-prop, because we kill sym-prop values at JOIN points, so any sym-prop values are lost after an if-statement. Sym-prop is experimental so that's OK.

Comment (IagoAbal) [reply cli/tests/e2e/snapshots/test_dependency_aware_rules/test_dependency_aware_rules/rulesdependency_awarelog4shell.yaml-dependency_awarelog4j/results.txt:42]: See also this test I added: https://github.com/returntocorp/semgrep/blob/38baaabc8d7d45ba2cf3d5c0e10ef2777c7a7826/tests/rules/sym_prop_no_merge4.java

File: src/analyzing/AST_to_IL.ml (hunk 2 above)
Comment (IagoAbal) [reply src/analyzing/AST_to_IL.ml:456]: Will add a comment.
